### PR TITLE
`odo dev`: Add support for non-exec commands part of a composite run or debug command

### LINKED ISSUE #5800: odo dev fails if multiple deployments exist for a component
Example devfile used for the project https://github.com/odo-devfiles/nodejs-ex:

<details>

<summary> devfile.yaml </summary>

```yaml
commands:
- exec:
    commandLine: npm install
    component: runtime
    group:
      isDefault: true
      kind: build
    workingDir: ${PROJECT_SOURCE}
  id: install
- exec:
    commandLine: npm start
    component: runtime
    group:
      isDefault: true
      kind: run
    workingDir: ${PROJECT_SOURCE}
  id: run
- exec:
    commandLine: npm run debug
    component: runtime
    group:
      isDefault: true
      kind: debug
    workingDir: ${PROJECT_SOURCE}
  id: debug
- exec:
    commandLine: npm test
    component: runtime
    group:
      isDefault: true
      kind: test
    workingDir: ${PROJECT_SOURCE}
  id: test
components:
- container:
    endpoints:
    - name: http-3000
      targetPort: 3000
    image: registry.access.redhat.com/ubi8/nodejs-14:latest
    memoryLimit: 1024Mi
    mountSources: true
    command: ["tail"]
    args: ["-f", "/dev/null"]
  name: runtime
- name: infinitepodtwo
  kubernetes:  
    inlined: |
      apiVersion: v1
      kind: Pod
      metadata:
        name: infinitepodtwo
      spec:
        containers:
          - name: blah
            image: docker.io/dharmit/infiniteloop
- name: infinitepodone
  kubernetes:  
    inlined: |
      apiVersion: v1
      kind: Pod
      metadata:
        name: infinitepodone
      spec:
        containers:
          - name: blah
            image: docker.io/dharmit/infiniteloop
- name: mydeployment
  kubernetes:
    inlined: |
      apiVersion: apps/v1
      kind: Deployment
      metadata:
        name: infinitedeployment
      spec:
        replicas: 1
        selector:
          matchLabels:
            app: infinite
        template:
          metadata:
            labels:
              app: infinite
          spec:
            containers:
            - name: blah
              image: docker.io/dharmit/infiniteloop
metadata:
  description: Stack with Node.js 14
  displayName: Node.js Runtime
  icon: https://nodejs.org/static/images/logos/nodejs-new-pantone-black.svg
  language: javascript
  name: node
  projectType: nodejs
  tags:
  - NodeJS
  - Express
  - ubi8
  version: 1.0.1
schemaVersion: 2.0.0
starterProjects:
- git:
    remotes:
      origin: https://github.com/odo-devfiles/nodejs-ex.git
  name: nodejs-starter
```

</details>

First pass of `odo dev` works just fine. Modify `server.js` and save the file. `odo dev` which is watching the current working directory fails with:

```sh
$ odo dev
  __
 /  \__     Developing using the node Devfile
 \__/  \    Namespace: myproject
 /  \__/    odo version: v3.0.0-alpha2
 \__/

↪ Deploying to the cluster in developer mode
 ✓  Searching resource in cluster 
 ✓  Creating kind Pod [13ms]
 ✓  Searching resource in cluster 
 ✓  Creating kind Pod [15ms]
 ✓  Searching resource in cluster 
 ✓  Creating kind Deployment [13ms]
 ✓  Waiting for Kubernetes resources [3s]
 ✓  Syncing files into the container [158ms]
 ✓  Executing install command "npm install" on container "runtime" [2s]

Your application is now running on the cluster
 - Forwarding from 127.0.0.1:40001 -> 3000

Watching for changes in the current directory /home/dshah/src/nodejs-ex
Press Ctrl+c to exit `odo dev` and delete resources from the cluster


File /home/dshah/src/nodejs-ex/server.js changed
Pushing files...

Error occurred on Push - watch command was unable to push component: failed to create the component: unable to determine if component node exists: multiple Deployments exist for the selector: app.kubernetes.io/instance=node,app.kubernetes.io/managed-by=odo,app.kubernetes.io/part-of=app,odo.dev/mode=Dev. Only one must be present
```

/kind bug
/area dev

### LINKED ISSUE #5832: `odo dev`: Add support for non-exec commands part of a composite `run` or `debug` command
## User Story
- As an odo user
- Now that I am able to run commands that are composite (thanks to #5054)
- I want to be able to include non-exec commands(apply command) as part of my composite run or debug command
- So that I can perform additional operations, like applying other component definitions (e.g., for building and pushing images, creating other Kubernetes resources, ...)

## Acceptance Criteria 
- [ ] `odo dev` should run all the commands part of the composite `run` command
- [ ] `odo dev --debug` should run all the commands part of the composite `debug` command

## Example

```yaml
schemaVersion: 2.2.0
metadata:
  name: nodejs
variables:
  CONTAINER_IMAGE: quay.io/unknown-account/myimage

components:
  - container:
      endpoints:
        - name: http-3000
          targetPort: 3000
      image: registry.access.redhat.com/ubi8/nodejs-14:latest
    name: runtime

  - name: image-build-component
    image:
      imageName: "{{CONTAINER_IMAGE}}"
      dockerfile:
        uri: ./Dockerfile

  - name: my-k8s-resource
    kubernetes:
      # Resource potentially referencing the 'CONTAINER_IMAGE' variable.
      # Might be inlined as well.
      uri: 'k8s/my-resource.yaml'

commands:
  - id: build-image
    apply:
      component: image-build-component

  - id: create-k8s-resource
    apply:
      component: my-k8s-resource

  - id: install
    exec:
      commandLine: npm install
      component: runtime
      workingDir: /projects

  - id: start
    exec:
      workingDir: /projects
      commandLine: npm start
      component: runtime

  - id: run
    composite:
      commands:
        - build-image
        - create-k8s-resource
        - install
        - start
      group:
        isDefault: true
        kind: run
```

## Links 
- Related Epic (mandatory): 
- Blocked by: https://github.com/redhat-developer/odo/issues/5054

/kind user-story

### PR DESCRIPTION
**Pull request by valaparthvi**
**What type of PR is this:**
/kind feature

**What does this PR do / why we need it:**

**Which issue(s) this PR fixes:**

Fixes #5832 
Fixes #5800

**PR acceptance criteria:**

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**
1. mkdir /tmp/111; cd /tmp/111
2. Use the devfile described in the issue or in this PR,
3. mkdir k8s
4. `kubectl create deploy odo-dev --image=nodejs -oyaml --dry-run=client > k8s/my-resource.yaml`
5. `PODMAN_CMD="echo"`
6. `odo dev`
7. `kubectl get deploy` should show `odo-dev` deployment.

### Problems with the current approach:

- [x] 1. If a deployment resource is created with composite-apply, it will lead to multiple deployments found error, because both the main (component) deployment, and the deployment resource have the same labels. https://github.com/redhat-developer/odo/pull/5923#issuecomment-1177286376
     **Solution:** This issue can be addressed with the label "component" that is only applied to the core resources created by `odo dev`. See - https://github.com/redhat-developer/odo/pull/5923/commits/94cb56b26387dd41b1a69df4d09d3eba3287b4ad
- [x] 2. `odo dev` does not clean up the resources created by apply command. https://github.com/redhat-developer/odo/pull/5923#issuecomment-1176018723
     **Solution:** This issue is addressed by modifying `ListResourcesToDeleteFromDevfile` method. This method previously assumed that `odo dev` only created deployment, which was correct, but with the changes in this PR, `odo dev` can now create a variety of resources by `apply` command. See - https://github.com/redhat-developer/odo/pull/5923/commits/e77df561980bddb09b9bed9a08e7ce5e97d474c8
- [x] 3. The code coincides with [deployHandler](https://github.com/redhat-developer/odo/blob/main/pkg/deploy/deploy.go#L73), it needs to be DRY-ed.

### CODE REVIEW

**Comment by valaparthvi** [comment]
@rm3l @feloy this is a temporary PR that I've created. It simply copy pastes logic from deployHandler's methods to the commonhandler.
It seems to be doing the trick, but I'd like to confirm if this is what we want before proceeding further.

**Comment by rm3l** [comment]
> this is a temporary PR that I've created. It simply copy pastes logic from deployHandler's methods to the commonhandler. It seems to be doing the trick, but I'd like to confirm if this is what we want before proceeding further and fixing problems with this implementation, for e.g. it does not cleanup the resource created by apply command on exiting odo dev.
>
Yes, after quickly testing this, I think this is what is expected here. Maybe we could mutualize that logic between both handlers..

There might be other side effects due to this.
For example, odo might no longer sync files if the composite command references a command applying a Kubernetes component, because it was finding multiple Deployments with the selector.
```
Error occurred on Push - watch command was unable to push component: 
unable to determine if component my-nodejs-app-with-composite-run-commands exists: 
multiple Deployments exist for the selector: 
  app.kubernetes.io/instance=my-nodejs-app-with-composite-run-commands,
  app.kubernetes.io/managed-by=odo,
  app.kubernetes.io/part-of=app,
  odo.dev/mode=Dev. 
Only one must be present
```

File: pkg/storage/kubernetes.go [hunk 1]
```diff
@@ -86,7 +86,9 @@ func (k kubernetesClient) Delete(name string) error {
 func (k kubernetesClient) List() (StorageList, error) {
 	if k.deployment == nil {
 		var err error
-		k.deployment, err = k.client.GetOneDeployment(k.componentName, k.appName)
+		// FIXME: Find a better way to add the component label to selector
+		selector := fmt.Sprintf("%s,component=%s", odolabels.GetSelector(k.componentName, k.appName, odolabels.ComponentAnyMode), k.componentName)
```
Comment (rm3l) [inline pkg/storage/kubernetes.go:90]: Same comment as before: Would a better way be to add a new boolean to the labels.GetSelector function to indicate whether it should include the component label as well?

**Review by rm3l — CHANGES_REQUESTED** [review]

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 2]
```diff
@@ -116,7 +116,9 @@ func (a Adapter) Push(parameters adapters.PushParameters) (err error) {
 	// Get the Dev deployment:
 	// Since `odo deploy` can theoretically deploy a deployment as well with the same instance name
 	// we make sure that we are retrieving the deployment with the Dev mode, NOT Deploy.
-	selectorLabels := odolabels.GetSelector(a.ComponentName, a.AppName, odolabels.ComponentDevMode)
+	// FIXME: Find a better way to add this label to selector
+	selectorLabels := fmt.Sprintf("%s,component=%s", odolabels.GetSelector(a.ComponentName, a.AppName, odolabels.ComponentDevMode), a.ComponentName)
```
[the remark this answers is not in the document]
Comment (feloy) [reply pkg/devfile/adapters/kubernetes/component/adapter.go:120]: Yes, and the Boolean name should indicate the intent (as I understand, the Deployment is part of the component), and not the implement details (component label is used).

To say it differently, you could add a Boolean `isPartOfComponent`, or similar.

File: pkg/component/delete/delete.go [hunk 3]
```diff
@@ -120,27 +121,25 @@ func (do DeleteComponentClient) ListResourcesToDeleteFromDevfile(devfileObj pars
...
-			cr, err = do.kubeClient.GetDynamicResource(gvr.Resource, lr.GetName())
-			if err != nil {
-				continue
-			}
-			resources = append(resources, *cr)
+		// Try to fetch the resource from the cluster; if it exists, append it to the resources list
+		var cr *unstructured.Unstructured
+		cr, err = do.kubeClient.GetDynamicResource(gvr.Resource, lr.GetName())
+		// If a specific mode is asked for, then make sure it matches with the cr's mode.
+		if err != nil || (mode != odolabels.ComponentAnyMode && odolabels.GetMode(cr.GetLabels()) != mode) {
```
[the remark this answers is not in the document]
Comment (valaparthvi) [reply pkg/component/delete/delete.go:139]: I've added a test for this. I had to modify the unit test for it, but there aren't any major changes other than adding mode arg, changing variable names, and adding 2 other unit tests.

**Comment by valaparthvi** [comment]
> > 5. PODMAN_CMD="echo"
> 
> What's the expectation if I don't specify this? Should `odo dev` build the container image using the Dockerfile in pwd? Currently, the code in this PR is not doing so. What's the purpose - from the user's perspective - to specify an `image` component in such a composite `run` command?

I asked to set PODMAN_CMD because the image used by the devfile does not point to the right image registry, if you in fact use a valid image registry, and have a Dockerfile, things should work as expected.. If you used some other devfile, please let me know, I can test it.

**Comment by valaparthvi** [comment]
Integration tests are failing on IBM cloud, but I ran them locally, and they pass.

**Review by rm3l — CHANGES_REQUESTED** [review]
LGTM feature-wise.

/approve

I just added a few comments in the test cases added.

Since this relates to `debug` commands as well, could you also add a test case in `cmd_dev_debug_test`?

File: tests/integration/cmd_dev_debug_test.go [hunk 4]
```diff
@@ -122,6 +122,72 @@ var _ = Describe("odo dev debug command tests", func() {
...
+		var sessionOut []byte
+		var err error
+		var ports map[string]string
+		const (
+			DEVFILE_DEBUG_PORT = "5858"
+		)
+
+		BeforeEach(func() {
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), commonVar.Context)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile-composite-apply-commands-debug.yaml"), filepath.Join(commonVar.Context, "devfile.yaml"))
```
Comment (rm3l) [inline tests/integration/cmd_dev_debug_test.go:137]: nit: `devfile-composite-apply-commands-debug.yaml` and `devfile-composite-apply-commands.yaml` look quite similar. Maybe you could just add the debug command (and endpoint) to `devfile-composite-apply-commands-debug.yaml` and use it here as well?

Comment (valaparthvi) [reply tests/integration/cmd_dev_debug_test.go:137]: Definitely not a nit! That's a good suggestion :smile:

**Comment by rm3l** [comment]
@valaparthvi Looks like some validation checks and some tests are not passing. Can you take a look at them?

File: tests/examples/source/devfiles/nodejs/devfile- [hunk 5]
```diff
@@ -0,0 +1,75 @@
+schemaVersion: 2.2.0
```
Comment (rm3l) [inline tests/examples/source/devfiles/nodejs/devfile-:1]: This file should be removed altogether.

File: tests/integration/cmd_dev_test.go [hunk 6]
```diff
@@ -1488,7 +1555,7 @@ var _ = Describe("odo dev command tests", func() {
 			checkFunc         func(stdout, stderr string)
 		}
 		testForCmd := func(tt testCase) {
-			err := helper.RunDevMode(tt.devAdditionalOpts, func(session *gexec.Session, outContents []byte, errContents []byte, ports map[string]string) {
+			err := helper.RunDevMode(nil, tt.devAdditionalOpts, func(session *gexec.Session, outContents []byte, errContents []byte, ports map[string]string) {
```
Comment (rm3l) [inline tests/integration/cmd_dev_test.go:1558]: I think the two arguments are inverted. That might explain why the corresponding tests are no longer passing.

```suggestion
			err := helper.RunDevMode(tt.devAdditionalOpts, nil, func(session *gexec.Session, outContents []byte, errContents []byte, ports map[string]string) {
```

**Review by rm3l — CHANGES_REQUESTED** [review]

**Comment by valaparthvi** [comment]
All the tests pass locally for me, I am not sure what the trouble is.
<details>
<summary>O/p</summary>

```console
GOROOT=/usr/local/go #gosetup
GOPATH=/home/pvala/go #gosetup
/usr/local/go/bin/go test -c -o /tmp/GoLand/___go_test_github_com_redhat_developer_odo_tests_integration.test github.com/redhat-developer/odo/tests/integration #gosetup
/usr/local/go/bin/go tool test2json -t /tmp/GoLand/___go_test_github_com_redhat_developer_odo_tests_integration.test -test.v -test.paniconexit0
=== RUN   TestIntegration
Running Suite: Integration Suite - /home/pvala/go/src/github.com/redhat-developer/odo/tests/integration
=======================================================================================================
Random Seed: 1659521891

Will run 8 of 288 specs
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [SLOW TEST] [38.356 seconds]
odo dev command tests running odo dev for a devfile with composite apply command using a Devfile with an image component that uses a remote Dockerfile when [PODMAN_CMD=echo] remote server returns an error should not build images when odo dev is run
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_dev_test.go:1290
------------------------------
S
------------------------------
• [SLOW TEST] [35.424 seconds]
odo dev command tests running odo dev for a devfile with composite apply command using a Devfile with an image component that uses a remote Dockerfile when [PODMAN_CMD=a-command-not-found-for-podman-should-make-odo-fallback-to-docker DOCKER_CMD=echo] remote server returns an error should not build images when odo dev is run
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_dev_test.go:1290
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS;5;14mSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS4mSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS[0mSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [SLOW TEST] [10.776 seconds]
odo logs command tests when component is created and odo logs is executed when running in Dev mode should successfully show logs of the running component
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_logs_test.go:87
------------------------------
• [SLOW TEST] [40.422 seconds]
odo logs command tests when component is created and odo logs is executed when running in Dev mode when --follow flag is specified should successfully follow logs of running component
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_logs_test.go:111
------------------------------
• [SLOW TEST] [12.882 seconds]
odo logs command tests when component is created and odo logs is executed when running in Deploy mode should successfully show logs of the running component
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_logs_test.go:132
------------------------------
• [SLOW TEST] [32.896 seconds]
odo logs command tests when component is created and odo logs is executed when running in Deploy mode when --follow flag is specified should successfully follow logs of running component
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_logs_test.go:156
------------------------------
• [SLOW TEST] [29.120 seconds]
odo logs command tests when component is created and odo logs is executed when running in both Dev and Deploy mode should successfully show logs of the running component
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_logs_test.go:185
------------------------------
• [SLOW TEST] [49.411 seconds]
odo logs command tests when component is created and odo logs is executed when running in both Dev and Deploy mode when --follow flag is specified should successfully follow logs of running component
/home/pvala/go/src/github.com/redhat-developer/odo/tests/integration/cmd_logs_test.go:213
------------------------------

Ran 8 of 288 Specs in 249.292 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 280 Skipped
PASS | FOCUSED


Process finished with the exit code 1

```
</details>

**Comment by dharmit** [comment]
@valaparthvi is it the IBM Cloud where `odo logs` tests are failing for you? And is it only for OCP or for k8s as well? And are they failing with a message like below?
```
✗  failed to get logs for container main; error: container "main" in pod "myapp" is waiting to start: trying and failing to pull image
```

**Comment by valaparthvi** [comment]
> @valaparthvi is it the IBM Cloud where `odo logs` tests are failing for you? And is it only for OCP or for k8s as well? And are they failing with a message like below?
> 
> ```
> ✗  failed to get logs for container main; error: container "main" in pod "myapp" is waiting to start: trying and failing to pull image
> ```

Not with that error, no. And only OCP IBM cloud tests are failing for odo logs.

**Comment by dharmit** [comment]
> Not with that error, no. And only OCP IBM cloud tests are failing for odo logs.

OK. I have been seeing `odo logs` failures on my PR as well. But don't understand why that's the case. Please share the logs if you see the failures next time.

**Review by rm3l — APPROVED** [review]
Thanks for your work on this!

File: tests/integration/cmd_dev_test.go [hunk 7]
```diff
@@ -1147,21 +1149,172 @@ var _ = Describe("odo dev command tests", func() {
...
+					var url string
+					var out, sessionErr string
+					BeforeEach(func() {
+						server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+							w.WriteHeader(http.StatusNotFound)
+						}))
+						url = server.URL
+
+						helper.ReplaceString(filepath.Join(commonVar.Context, "devfile.yaml"), "./Dockerfile", url)
+						out, sessionErr = helper.Cmd("odo", "dev", "--random-ports").ShouldFail().OutAndErr()
```
Comment (feloy) [inline tests/integration/cmd_dev_test.go:1286]: You are not passing `env` here

**Comment by valaparthvi** [comment]
/override ci/prow/v4.10-integration-e2e
IBM tests pass.

**Pull request merged**